### PR TITLE
Allow creating indices when schema name != generated class name

### LINKED ISSUE #427: SchemaNotFoundException when using UniqueKeyIndex
### When hollow-generated class' name is postfixed, UniqueKeyIndex cannot find HollowSchema.  

Example of HollowAPIGenerator creation:
`  new HollowAPIGenerator.Builder().withClassPostfix("Hollow")...
`

In this case, the name of hollow-generated class from https://github.com/Netflix/hollow/releases/tag/v4.1.0 example would be _MovieHollow_.
Schema name would still be just _Movie_. 

In this case, UniqueKeyIndex **cannot** be built using this piece of code:
`UniqueKeyIndex.from(consumer, Movie.class).bindToPrimaryKey()...`

This is caused by _bindToPrimaryKey_ method implementation, which seeks for HollowSchema using _uniqueType_'s simple name, which in this case is _MovieHollow_, whereas schema name is _Movie_.

Mentioned method seeks for HollowSchema using the following piece of code:

` String primaryTypeName = uniqueType.getSimpleName();
HollowSchema schema = consumer.getStateEngine().getNonNullSchema(primaryTypeName);
`

### PR DESCRIPTION
**Pull request by akhaku**
- generate classes with @HollowTypeName if name does not match schema
- read the schema name using any @HollowTypeName when inspecting types during index creation

Fixes #427.

### CODE REVIEW

File: hollow/src/test/java/com/netflix/hollow/api/codegen/AbstractHollowAPIGeneratorTest.java [hunk 1]
```diff
@@ -55,7 +61,15 @@ void assertNonEmptyFileExists(String relativePath) throws IOException {
         assertTrue("File at " + relativePath + " should exist", f.exists() && f.length() > 0L);
     }
 
-    void assertFileDoesNotExist(String relativePath) throws IOException {
+    void assertClassHasHollowTypeName(String clazz, String typeName) throws IOException, ClassNotFoundException {
```
Comment (akhaku) [inline hollow/src/test/java/com/netflix/hollow/api/codegen/AbstractHollowAPIGeneratorTest.java:64]: Couldn't find a good way to test what we really want (that we can create generated indices) so went with the best I could - verify that a HollowTypeName is set.

File: hollow/src/main/java/com/netflix/hollow/api/codegen/objects/HollowObjectJavaGenerator.java [hunk 2]
```diff
@@ -105,6 +106,9 @@ public String generate() {
         classBuilder.append("\n");
 
         classBuilder.append("@SuppressWarnings(\"all\")\n");
+        if (!className.equals(schema.getName())) {
+            classBuilder.append("@" + HollowTypeName.class.getName() + "(name=\"" + schema.getName() + "\")\n");
```
Comment (PaulSandoz) [inline hollow/src/main/java/com/netflix/hollow/api/codegen/objects/HollowObjectJavaGenerator.java]: That's gonna use the fully qualified name.  Can you declare an import?

File: hollow/src/test/java/com/netflix/hollow/api/codegen/AbstractHollowAPIGeneratorTest.java (hunk 1 above)
Comment (PaulSandoz) [reply hollow/src/test/java/com/netflix/hollow/api/codegen/AbstractHollowAPIGeneratorTest.java:64]: That seems reasonable.  We can work on an additional fix to the indexes.  If you look in the test class `com.netflix.hollow.api.consumer.index.DataModel.Consumer` we can add something explicit.

File: hollow/src/main/java/com/netflix/hollow/api/codegen/objects/HollowObjectJavaGenerator.java (hunk 2 above)
Comment (akhaku) [reply hollow/src/main/java/com/netflix/hollow/api/codegen/objects/HollowObjectJavaGenerator.java]: Sure, this was the lazy way to go

File: hollow/src/test/java/com/netflix/hollow/api/codegen/AbstractHollowAPIGeneratorTest.java (hunk 1 above)
Comment (akhaku) [reply hollow/src/test/java/com/netflix/hollow/api/codegen/AbstractHollowAPIGeneratorTest.java:64]: I don't think I understand your comment - is there a way to test the generated index, or are you suggesting adding a test to make sure that the HollowTypeName is picked up by indexes?

File: hollow/src/main/java/com/netflix/hollow/api/consumer/index/SelectFieldPathResultExtractor.java [hunk 3]
```diff
@@ -71,7 +72,7 @@ static IllegalArgumentException incompatibleSelectType(Class<?> selectType, Stri
     static <T> SelectFieldPathResultExtractor<T> from(
             Class<? extends HollowAPI> apiType, HollowDataset dataset, Class<?> rootType, String fieldPath,
             Class<T> selectType) {
-        String rootTypeName = rootType.getSimpleName();
+        String rootTypeName = HollowObjectTypeMapper.getDefaultTypeName(rootType);
```
Comment (akhaku) [inline hollow/src/main/java/com/netflix/hollow/api/consumer/index/SelectFieldPathResultExtractor.java:75]: I had some trouble testing the changes in this file, Paul's eyes would be useful here

Comment (PaulSandoz) [reply hollow/src/main/java/com/netflix/hollow/api/consumer/index/SelectFieldPathResultExtractor.java:75]: Here's an example:
```
HashIndexSelect<Movie, Actor, String> index = HashIndex.from(consumer, Movie.class)
    .selectField("cast.element.actor", Actor.class)
    .usingPath("title.value", String.class);

String title = "The Imitation Game";
Stream<Actor> m = index.findMatch(title);  
```
It's a bit like:
```
SELECT Actor FROM Movie
WHERE title='The Imitation Game';
```
So you need to add another annotated test class for selection type referenced from the top type.

File: hollow/src/test/java/com/netflix/hollow/api/consumer/index/HashIndexTest.java [hunk 4]
```diff
@@ -57,7 +58,11 @@ public static void setup() {
 
             for (int i = 0; i < 100; i++) {
                 ws.add(new DataModel.Producer.TypeA(1, "TypeA" + i));
+                ws.add(new DataModel.Producer.TypeB(1, "TypeB" + i));
```
Comment (akhaku) [inline hollow/src/test/java/com/netflix/hollow/api/consumer/index/HashIndexTest.java:61]: Something odd is going  on - somehow adding this here causes TestManyMatches#testWithSelectGenericHollowObject,testWithSelect to fail. Digging further...

Comment (akhaku) [reply hollow/src/test/java/com/netflix/hollow/api/consumer/index/HashIndexTest.java:61]: I can reproduce using an old-style manually constructed HollowHashIndex. It only happens when I pass in a select field path for the string and I _don't_ explicitly as for `.value` as the select field path. This is explicitly disallowed in the new-style hash indexes - need to sync up with Paul on this.

Comment (PaulSandoz) [reply hollow/src/test/java/com/netflix/hollow/api/consumer/index/HashIndexTest.java:61]: The field path parsing should be consistent with `com.netflix.hollow.core.index.HollowHashIndex` and `com.netflix.hollow.api.consumer.index.HashIndex`. See use of `FieldPaths.createFieldPathForHashIndex`.  Auto-expansion is not supported for the select and match field paths of a hash index.

File: hollow/src/test/java/com/netflix/hollow/api/codegen/AbstractHollowAPIGeneratorTest.java (hunk 1 above)
Comment (PaulSandoz) [reply hollow/src/test/java/com/netflix/hollow/api/codegen/AbstractHollowAPIGeneratorTest.java]: I found the cause.  Select path never quite worked correctly, it currently requires the path have a last segment reference a field value, and the ordinal returned will correspond to the enclosing record.

The problematic line is at https://github.com/Netflix/hollow/blob/master/hollow/src/main/java/com/netflix/hollow/core/index/traversal/TraversalTreeBuilder.java#L103:
```java
                            if(schema.getFieldType(fieldSchemaPositions[i]) == FieldType.REFERENCE) {
                                fieldSchemaPositions[i] = -1;
                                fieldTypeDataAccess[i] = getChildDataAccess(typeDataAccess, pathElement);
                            } else {
                                fieldSchemaPositions[i] = schema.getPosition(pathElement);
                                fieldTypeDataAccess[i] = typeDataAccess;
                            }
```

It should be:
```java
                            if(schema.getFieldType(pathElement) == FieldType.REFERENCE) {
```

File: hollow/src/test/java/com/netflix/hollow/api/consumer/index/HashIndexTest.java (hunk 4 above)
Comment (PaulSandoz) [reply hollow/src/test/java/com/netflix/hollow/api/consumer/index/HashIndexTest.java:61]: I found the cause.  Select path never quite worked correctly, it currently requires the path have a last segment reference a field value, and the ordinal returned will correspond to the enclosing record.

The problematic line is at https://github.com/Netflix/hollow/blob/master/hollow/src/main/java/com/netflix/hollow/core/index/traversal/TraversalTreeBuilder.java#L103:
```java
                            if(schema.getFieldType(fieldSchemaPositions[i]) == FieldType.REFERENCE) {
                                fieldSchemaPositions[i] = -1;
                                fieldTypeDataAccess[i] = getChildDataAccess(typeDataAccess, pathElement);
                            } else {
                                fieldSchemaPositions[i] = schema.getPosition(pathElement);
                                fieldTypeDataAccess[i] = typeDataAccess;
                            }
```

It should be:
```java
                            if(schema.getFieldType(pathElement) == FieldType.REFERENCE) {
```

Comment (akhaku) [reply hollow/src/test/java/com/netflix/hollow/api/consumer/index/HashIndexTest.java:61]: nice - great find!

**Review by PaulSandoz — APPROVED** [review]

**Pull request merged**